File: data-export/ink-density-tool/export/excel.py
"""Excel export via openpyxl.

Template structure (DGC-curve-calculator.xlsx) — confirmed from file inspection:

  4 sheets in 2 pairs, one pair per dot shape:
    Pair 0 (shapes[0]): Sheet1 = weight[0] (120#), Sheet2 = weight[1] (150#) + weight[2] (200#)
    Pair 1 (shapes[1]): Sheet3 = weight[0],        Sheet4 = weight[1] + weight[2]

  Single-weight sheet layout (Sheet1 / Sheet3):
    A1      : job title  "{customer} {stock}"
    I1      : date
    B4:E17  : CMYK step readings for weight[0], steps 100→1  (14 rows × 4 cols)
              or B4:E19 for 16-step jobs (adds 0.8%, 0.4%)
    A20     : weight[0].label  (row = 4 + num_steps + 2)
    I20     : job.crs

  Dual-weight sheet layout (Sheet2 / Sheet4):
    A1      : job title
    I1      : date
    B4:E17  : CMYK step readings for weight[1]
    A20     : weight[1].label
    I20     : job.crs
    B27:E40 : CMYK step readings for weight[2]
    A43     : weight[2].label
    I43     : job.crs

  Note: rows 18–19 (0.8%, 0.4% steps) and rows 41–42 are filled for 16-step jobs.

  Bug fix: In Sheet2/Sheet4 the second table's F-column formulas (F27:F40) reference
  wrong rows due to a copy-paste error in the original template.  We rewrite them with
  the correct row references when exporting.

  The density readings (2.11, 1.80, etc.) are written to the Illustrator PDF only;
  the Excel template has no density row.
"""
from __future__ import annotations

import sys
from pathlib import Path

import openpyxl
from openpyxl.cell import MergedCell
from openpyxl.worksheet.page import PageMargins
from openpyxl.worksheet.properties import WorksheetProperties, PageSetupProperties

from core.models import JobConfig, ShapeData


# Row where step data starts (first table, both sheet types)
_STEP_START_ROW_T1 = 4
# Gap (in rows) between the T1 label row and the T2 step-start row
_GAP_T1_TO_T2 = 7
# CMYK data columns: B=2, C=3, D=4, E=5
_DATA_COLS = [2, 3, 4, 5]


def _row_constants(num_steps: int) -> tuple[int, int, int]:
    """Return (label_t1, step_start_t2, label_t2) for a given step count.

    Derivation (verified against both bundled templates):
      label_t1      = STEP_START_T1 + num_steps          (label immediately follows last step)
      step_start_t2 = label_t1 + GAP_T1_TO_T2
      label_t2      = step_start_t2 + num_steps
    """
    label_t1 = _STEP_START_ROW_T1 + num_steps
    step_start_t2 = label_t1 + _GAP_T1_TO_T2
    label_t2 = step_start_t2 + num_steps
    return label_t1, step_start_t2, label_t2


def _bundled_template(num_steps: int) -> Path:
    """Return the path to the bundled template matching the step count."""
    base = Path(sys._MEIPASS) if getattr(sys, "frozen", False) else Path(__file__).parent.parent
    name = "template_extended.xlsx" if num_steps > 14 else "template_standard.xlsx"
    return base / "assets" / name


def export_excel(job: JobConfig, output_path: str) -> None:
    """Fill the Excel template from job data and save to output_path."""
    num_steps = len(job.step_labels)

    # Use explicit override path if set, otherwise auto-select bundled template
    if job.template_xlsx_path:
        template_path = Path(job.template_xlsx_path)
    else:
        template_path = _bundled_template(num_steps)

    if not template_path.is_file():
        raise FileNotFoundError(f"Excel template not found: {template_path}")

    wb = openpyxl.load_workbook(str(template_path))
    label_t1, step_start_t2, label_t2 = _row_constants(num_steps)

    title = " ".join(filter(None, [
        job.customer, job.print_type, job.stock_desc, job.finish
    ]))
    dot_shape = " ".join(filter(None, [job.dot_shape_type, job.dot_shape_number]))

    for shape_idx, shape in enumerate(job.shapes):
        # Each shape occupies a pair of sheets (indices 2*shape_idx and 2*shape_idx + 1)
        single_idx = shape_idx * 2      # Sheet1, Sheet3, Sheet5…
        dual_idx   = shape_idx * 2 + 1  # Sheet2, Sheet4, Sheet6…

        # If the template doesn't have enough sheets, add new ones by copying
        while len(wb.sheetnames) <= dual_idx:
            _append_sheet_pair(wb, num_steps)

        ws_single = wb[wb.sheetnames[single_idx]]
        ws_dual   = wb[wb.sheetnames[dual_idx]]

        # --- Single sheet: weight[0] ---
        _write_metadata(ws_single, title, job.date)
        if len(shape.weights) > 0:
            w0 = shape.weights[0]
            _write_steps(ws_single, w0.steps, _STEP_START_ROW_T1, num_steps)
            _write_cell(ws_single, label_t1, 1, w0.label)
            _write_cell(ws_single, label_t1, 9, dot_shape)

        # --- Dual sheet: weight[1] (first table) ---
        _write_metadata(ws_dual, title, job.date)
        if len(shape.weights) > 1:
            w1 = shape.weights[1]
            _write_steps(ws_dual, w1.steps, _STEP_START_ROW_T1, num_steps)
            _write_cell(ws_dual, label_t1, 1, w1.label)
            _write_cell(ws_dual, label_t1, 9, dot_shape)

        # --- Dual sheet: weight[2] (second table) ---
        if len(shape.weights) > 2:
            w2 = shape.weights[2]
            _write_steps(ws_dual, w2.steps, step_start_t2, num_steps)
            _write_cell(ws_dual, label_t2, 1, w2.label)
            _write_cell(ws_dual, label_t2, 9, dot_shape)
            _fix_second_table_formulas(ws_dual, step_start_t2, num_steps)

    # Apply A4 page setup to all worksheets
    for ws in wb.worksheets:
        ws.page_setup.paperSize = 9           # 9 = A4
        ws.page_setup.orientation = "portrait"
        ws.page_setup.fitToWidth = 1
        ws.page_setup.fitToHeight = 0         # auto height
        # fitToPage must be set via sheet_properties to avoid NoneType error
        # on worksheets loaded from existing files
        if ws.sheet_properties is None:
            ws.sheet_properties = WorksheetProperties()
        ws.sheet_properties.pageSetUpPr = PageSetupProperties(fitToPage=True)
        ws.print_area = f"A1:I{ws.max_row}"
        ws.page_margins = PageMargins(
            left=0.5, right=0.5, top=0.75, bottom=0.75,
            header=0.3, footer=0.3,
        )

    Path(output_path).parent.mkdir(parents=True, exist_ok=True)
    wb.save(output_path)


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _write_cell(ws, row: int, col: int, value) -> None:
    """Write value to a cell, resolving to the top-left cell if it is part of a merge."""
    cell = ws.cell(row=row, column=col)
    if isinstance(cell, MergedCell):
        for rng in ws.merged_cells.ranges:
            if rng.min_row <= row <= rng.max_row and rng.min_col <= col <= rng.max_col:
                ws.cell(row=rng.min_row, column=rng.min_col, value=value)
                return
    else:
        cell.value = value


def _write_metadata(ws, title: str, date: str) -> None:
    _write_cell(ws, 1, 1, title)   # A1
    _write_cell(ws, 1, 9, date)    # I1


def _write_steps(ws, steps: list[list[float]], start_row: int, num_steps: int) -> None:
    """Write num_steps rows of CMYK step readings into columns B–E starting at start_row."""
    for ri in range(num_steps):
        if ri >= len(steps):
            break
        row_data = steps[ri]
        excel_row = start_row + ri
        for ci, col in enumerate(_DATA_COLS):
            value = row_data[ci] if ci < len(row_data) else None
            # Write blank rather than 0 for unset values so formulas read as empty
            _write_cell(ws, excel_row, col, value if value else None)


def _fix_second_table_formulas(ws, start_row: int, num_steps: int) -> None:
    """Correct the Average% formulas in the second table's F column.

    The original template has a copy-paste error where F27 references B25:E25
    instead of B27:E27 (offset by 2 rows).  We overwrite F{start_row}:F{end_row}
    with correct =SUM(B{r}:E{r})/4 formulas.
    """
    for i in range(num_steps):
        r = start_row + i
        ws.cell(row=r, column=6, value=f"=SUM(B{r}:E{r})/4")


def _append_sheet_pair(wb: openpyxl.Workbook, num_steps: int) -> None:
    """Add a blank single+dual sheet pair by copying Sheet1/Sheet2 structure."""
    existing = wb.sheetnames
    src_single = wb[existing[0]]
    src_dual   = wb[existing[1]]

    new_single = wb.copy_worksheet(src_single)
    new_dual   = wb.copy_worksheet(src_dual)

    _, step_start_t2, _ = _row_constants(num_steps)
    t1_clear_end = _STEP_START_ROW_T1 + num_steps + 1
    t2_clear_end = step_start_t2 + num_steps + 1

    for ws in (new_single, new_dual):
        for row in range(_STEP_START_ROW_T1, t1_clear_end):
            for col in range(2, 6):
                ws.cell(row=row, column=col, value=None)
    for row in range(step_start_t2, t2_clear_end):
        for col in range(2, 6):
            new_dual.cell(row=row, column=col, value=None)

File: data-export/ink-density-tool-rs/src/export/excel.rs
//! Excel export using rust_xlsxwriter.
//!
//! Since rust_xlsxwriter creates new workbooks (can't modify templates like openpyxl),
//! we recreate the template structure programmatically. The layout matches the Python
//! export exactly.

use std::path::Path;

use anyhow::{Context, Result};
use rust_xlsxwriter::{Format, Workbook, Worksheet};

use crate::core::models::JobConfig;

/// Row where step data starts (first table, both sheet types)
const STEP_START_ROW_T1: u32 = 3; // 0-indexed (row 4 in Excel)
/// Gap (in rows) between the T1 label row and the T2 step-start row
const GAP_T1_TO_T2: u32 = 7;
/// CMYK data columns: B=1, C=2, D=3, E=4 (0-indexed)
const DATA_COLS: [u16; 4] = [1, 2, 3, 4];

fn row_constants(num_steps: u32) -> (u32, u32, u32) {
    let label_t1 = STEP_START_ROW_T1 + num_steps;
    let step_start_t2 = label_t1 + GAP_T1_TO_T2;
    let label_t2 = step_start_t2 + num_steps;
    (label_t1, step_start_t2, label_t2)
}

pub fn export_excel(job: &JobConfig, output_path: &Path) -> Result<()> {
    let num_steps = job.num_steps() as u32;
    let (label_t1, step_start_t2, label_t2) = row_constants(num_steps);

    let title = job.heading();
    let dot_shape = job.dot_shape();

    let mut workbook = Workbook::new();

    let header_fmt = Format::new().set_bold();
    let step_labels = &job.step_labels;

    for (shape_idx, shape) in job.shapes.iter().enumerate() {
        // --- Single-weight sheet: weight[0] ---
        let ws_name = if shape_idx == 0 {
            "Sheet1".to_string()
        } else {
            format!("Sheet{}", shape_idx * 2 + 1)
        };

        let ws = workbook.add_worksheet();
        ws.set_name(&ws_name)?;

        // Metadata
        ws.write_string_with_format(0, 0, &title, &header_fmt)?;
        ws.write_string(0, 8, &job.date)?;

        // Step labels in column A
        for (i, label) in step_labels.iter().enumerate() {
            ws.write_string(STEP_START_ROW_T1 + i as u32, 0, label)?;
        }

        if !shape.weights.is_empty() {
            let w0 = &shape.weights[0];
            write_steps(ws, &w0.steps, STEP_START_ROW_T1, num_steps)?;
            ws.write_string(label_t1, 0, &w0.label)?;
            ws.write_string(label_t1, 8, &dot_shape)?;
        }

        // --- Dual-weight sheet: weight[1] + weight[2] ---
        let ws_name2 = if shape_idx == 0 {
            "Sheet2".to_string()
        } else {
            format!("Sheet{}", shape_idx * 2 + 2)
        };

        let ws2 = workbook.add_worksheet();
        ws2.set_name(&ws_name2)?;

        ws2.write_string_with_format(0, 0, &title, &header_fmt)?;
        ws2.write_string(0, 8, &job.date)?;

        // Step labels for first table
        for (i, label) in step_labels.iter().enumerate() {
            ws2.write_string(STEP_START_ROW_T1 + i as u32, 0, label)?;
        }

        // weight[1] — first table
        if shape.weights.len() > 1 {
            let w1 = &shape.weights[1];
            write_steps(ws2, &w1.steps, STEP_START_ROW_T1, num_steps)?;
            ws2.write_string(label_t1, 0, &w1.label)?;
            ws2.write_string(label_t1, 8, &dot_shape)?;
        }

        // Step labels for second table
        for (i, label) in step_labels.iter().enumerate() {
            ws2.write_string(step_start_t2 + i as u32, 0, label)?;
        }

        // weight[2] — second table
        if shape.weights.len() > 2 {
            let w2 = &shape.weights[2];
            write_steps(ws2, &w2.steps, step_start_t2, num_steps)?;
            ws2.write_string(label_t2, 0, &w2.label)?;
            ws2.write_string(label_t2, 8, &dot_shape)?;

            // Fix second table formulas (F column)
            for i in 0..num_steps {
                let r = step_start_t2 + i;
                // 0-indexed: F = col 5, B = col 1, E = col 4
                let formula = format!(
                    "=SUM(B{}:E{})/4",
                    r + 1, // Excel 1-indexed for formula references
                    r + 1
                );
                ws2.write_formula(r, 5, rust_xlsxwriter::Formula::new(formula))?;
            }
        }
    }

    // Set up page layout for all sheets (A4, portrait, fit-to-width)
    // Note: rust_xlsxwriter handles this per-worksheet during creation

    if let Some(parent) = output_path.parent() {
        std::fs::create_dir_all(parent)
            .with_context(|| format!("Failed to create dir: {}", parent.display()))?;
    }

    workbook
        .save(output_path)
        .with_context(|| format!("Failed to save Excel: {}", output_path.display()))?;

    Ok(())
}

fn write_steps(
    ws: &mut Worksheet,
    steps: &[[f64; 4]],
    start_row: u32,
    num_steps: u32,
) -> Result<()> {
    for ri in 0..num_steps as usize {
        if ri >= steps.len() {
            break;
        }
        let row_data = &steps[ri];
        let excel_row = start_row + ri as u32;
        for (ci, &col) in DATA_COLS.iter().enumerate() {
            let value = if ci < 4 { row_data[ci] } else { 0.0 };
            if value != 0.0 {
                ws.write_number(excel_row, col, value)?;
            }
            // Write nothing (blank) for zero values so formulas read as empty
        }
    }
    Ok(())
}
